### PR TITLE
Be able to retrieve the authorization code from a Tuleap server

### PR DESCRIPTION
**Pull request by robinsoc**
This change is part of [community #14019](https://tuleap.net/plugins/tracker/?aid=14019)

To minimize the size of the pull request review, only the first step of the OAuth2 Authorization Code Grant[0] request[1] flow has been implemented.
So, the access token will be retrieved in a dedicated contribution.
In this contribution the user should be able to retrieve the authoriation code from a Tuleap server by filling the form in the Global Security menu in Jenkins.
When Jenkins get the authorization code, the user will be automatically logged as "Admin" in Jenkins to avoid to be closed outside of Jenkins during tests.

How to test:
 - Create a new Oauth2 app in Tuleap.
 - Go to "Configure Global Security" menu of Jenkins
 - Choose "Tuleap Authentication", fill the form and save.
 - Log out and log in again
=> You should be redirected to the Tuleap log in page. (if you are not already logged)
 - When logged, you should be redirected to the Athorization page of Tuleap.
=> You should be redirected to Jenkins whith the authorization code (see the web console) and you should be logged as Admin.
=> If there is an error, you should be redirected to a dedicated page.

[0] https://tools.ietf.org/html/rfc6749#section-4.1
[1] https://tools.ietf.org/html/rfc6749#section-4.1.1

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapLogoutAction.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    static final String REDIRECT_ON_LOGOUT = "tuleapLogout";
+    private PluginHelper pluginHelper;
+
+    public TuleapLogoutAction(){
+        Injector injector = Guice.createInjector(new TuleapOAuth2GuiceModule());
+        this.pluginHelper = injector.getInstance(PluginHelper.class);
+    }
+
+    // For testing purpose
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapLogoutAction.java]: why is it so ?

File: src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        }
+        return req.getContextPath() + "/" + TuleapLogoutAction.REDIRECT_ON_LOGOUT + "/";
+    }
+
+    public HttpResponse doCommenceLogin(StaplerRequest request) throws UnsupportedEncodingException {
+
+        final String state = this.buildStateRandomString();
+        request.getSession().setAttribute(STATE_SESSION_ATTRIBUTE, state);
+
+        final String scope = "read:project";
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java]: can we extract this scope to a constant

File: src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            "&redirect_uri=" + redirectUri +
+            "&scope=" + scope +
+            "&state=" + state
+        );
+    }
+
+    private String buildStateRandomString() {
+        final SecureRandom secureRandom = new SecureRandom();
+        final StringBuilder stateStringBuilder = new StringBuilder();
+        final String allowedCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~";
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java]: Can we extract this to a constant

File: src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java (hunk 2 above)
Comment (robinsoc) [reply src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java]: done

File: src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java (hunk 3 above)
Comment (robinsoc) [reply src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java]: done

**Comment by Erwyn** [comment]
Can we add a check for the trailing `/` in the Tuleap server adress ?

File: src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    @DataBoundConstructor
+    public TuleapSecurityRealm(String tuleapUri, String clientId, String clientSecret) {
+        this.setTuleapUri(Util.fixEmptyAndTrim(tuleapUri));
+        this.clientId = Util.fixEmptyAndTrim(clientId);
+        this.setClientSecret(Util.fixEmptyAndTrim(clientSecret));
+    }
+
+    public TuleapSecurityRealm() {
+
```
Comment (Erwyn) [inline src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java]: Can we remove this ?

Comment (robinsoc) [reply src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java]: done

**Pull request merged**